[Merged by Bors] - Move dependencies to workspace

jxs opened the pull request:
## Issue Addressed

Synchronize dependencies and edition on the workspace `Cargo.toml`

## Proposed Changes

with https://github.com/rust-lang/cargo/issues/8415 merged it's now possible to synchronize details on the workspace `Cargo.toml` like the metadata and dependencies.
By only having dependencies that are shared between multiple crates aligned on the workspace `Cargo.toml` it's easier to not miss duplicate versions of the same dependency and therefore ease on the compile times.

## Additional Info
this PR also removes the no longer required direct dependency of the `serde_derive` crate.

should be reviewed after https://github.com/sigp/lighthouse/pull/4639 get's merged.
closes https://github.com/sigp/lighthouse/issues/4651

jxs committed:
update dependencies

jxs committed:
switch procinfo for procfs,

to remove nom warning

jxs committed:
fix http and metrics flags

on beacon_node and validator_client tests
- fix tls test by adding cert and key file

jxs committed:
fix warp cors to allow all origins

on upstream warp

jxs committed:
move dependencies to workspace

jxs committed:
synchronize edition on the workspace Cargo.toml

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into move-to-workspace

jxs committed:
update Cargo.lock

jxs committed:
update Cargo.lock

jxs committed:
align rand crate, fix ef-tests

jxs committed:
revert zip update to reduce number of dependencies

jxs committed:
remove dangling comment

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into update-deps

jxs committed:
Merge branch 'update-deps' of github.com:jxs/lighthouse into move-to-workspace

jxs committed:
revert zip update to reduce number of dependencies

jxs committed:
Revert "revert zip update to reduce number of dependencies"

This reverts commit 281a26e6b2c78a05772d090a6c42cfc504fc1f0d.

jxs committed:
update libp2p to latest 0.52.3

jxs committed:
re-ignore RUSTSEC-2023-0052

jxs committed:
Merge branch 'update-deps' of github.com:jxs/lighthouse into move-to-workspace

jxs committed:
update deps to fix RUSTSEC-2023-0052

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into move-to-workspace

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into move-to-workspace

jxs committed:
fix quickcheck tests

jimmygchen commented on Cargo.toml:
```diff
@@ -89,10 +89,138 @@ members = [
+task_executor = { path = "common/task_executor" }
+types = { path = "consensus/types" }
+unused_port = { path = "common/unused_port" }
+validator_client = { path = "validator_client/" }
```
nitpick:

```suggestion
validator_client = { path = "validator_client" }
```

jxs replied:
thanks Jimmy!

jimmygchen reviewed (approved):
> this PR also removes the no longer required serde_derive crate.
What do you mean by this? `serde_derive` is still used and looks like this is one of the few crates that wasn't updated to workspace dependency. I'm curious about the reason behind this?

This is really nice! I'm not super familiar with cargo dependency management, but using workspace dependencies looks really neat, and I noticed a quite few duplicate dependencies have been removed which gives us more consistency in dependency versions and makes it a lot easier to maintain. 

I don't really have any other comments, the looks really good and tests are all passing so :clap:

jxs commented:
> > this PR also removes the no longer required serde_derive crate.
> > What do you mean by this? `serde_derive` is still used and looks like this is one of the few crates that wasn't updated to workspace dependency. I'm curious about the reason behind this?
> 
> This is really nice! I'm not super familiar with cargo dependency management, but using workspace dependencies looks really neat, and I noticed a quite few duplicate dependencies have been removed which gives us more consistency in dependency versions and makes it a lot easier to maintain.
> 
> I don't really have any other comments, the looks really good and tests are all passing so 👏

yeah sorry if it was misleading, what I meant regarding `serde-derive` is that we wouldn't be required to depend directly on it anymore. Updated the description to address this

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into move-to-workspace

jxs committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into move-to-workspace

jxs committed:
remove unused port usages

jxs committed:
update quinn proto, fix cargo-audit

paulhauner reviewed (commented):
Looks good to me, big effort!

Some of the crates in this repo are [published to crates.io](https://crates.io/teams/github:sigp:crates-io), I wonder if they're still publishable with `workspace = true` rather than the full dependency definition? I'm happy to approve and merge if they can still be published!

michaelsproul commented:
> Some of the crates in this repo are [published to crates.io](https://crates.io/teams/github:sigp:crates-io)

Are they? I think we already moved all of them out? All of `enr`, `discv5`, `ssz_types`, `ethereum_ssz`, `tree_hash`, `ethereum_hashing`, etc are in external repos:

- https://github.com/sigp/enr
- https://github.com/sigp/discv5
- https://github.com/sigp/ssz_types
- https://github.com/sigp/ethereum_ssz
- https://github.com/sigp/tree_hash
- https://github.com/sigp/ethereum_hashing

michaelsproul commented on Cargo.toml:
```diff
@@ -89,10 +89,138 @@ members = [
+rand = "0.8"
+rayon = "1.7"
+regex = "1"
+reqwest = { version = "0.11", features = ["blocking", "json", "stream", "native-tls-vendored"] }
```
We can (and should) turn off the `blocking` feature now that https://github.com/sigp/lighthouse/pull/4753 is merged

jxs replied:
yeah, thanks Michael, I was waiting for https://github.com/ralexstokes/ethereum-consensus/pull/261 to then switch `native-tls-vendored` for just `rustls-tls` but we can do that on a subsequent PR. Updated!

jxs replied:
https://github.com/sigp/lighthouse/actions/runs/6268152351/job/17022546605?pr=4650#step:8:54 seems to fail without `blocking`, reverted. I can address it in a subsequent PR so as to not block this one

michaelsproul replied:
oh yeah no worries!

jxs committed:
remove non required features from deps

jxs commented:
> Looks good to me, big effort!
> 
> Some of the crates in this repo are [published to crates.io](https://crates.io/teams/github:sigp:crates-io), I wonder if they're still publishable with `workspace = true` rather than the full dependency definition? I'm happy to approve and merge if they can still be published!

Thanks Paul, yeah afaik it's possible, [`rust-libp2p`](https://github.com/libp2p/rust-libp2p/) has the same architecture and publishes the subcrates.

jxs committed:
revert removal of blocking feature

michaelsproul commented on Cargo.toml:
```diff
@@ -89,10 +89,138 @@ members = [
+rand = "0.8"
+rayon = "1.7"
+regex = "1"
+reqwest = { version = "0.11", fault-features = false, features = ["blocking", "json", "stream", "rustls-tls"] }
```
```suggestion
reqwest = { version = "0.11", default-features = false, features = ["blocking", "json", "stream", "rustls-tls"] }
```

jimmygchen replied:
nice catch, I missed this one 😱

michaelsproul replied:
it just got added in the last few hours

michaelsproul committed:
Fix default-features typo

michaelsproul reviewed (approved):
Let's merge!!

michaelsproul commented:
Ah, there's a tokio dep issue introduced by 893b3eb6639e67ff7657d0b95a1c2dec6549f6a2. I'll fix that, test locally, and push

michaelsproul committed:
Add Tokio sync feature

paulhauner commented:
bors r+

michaelsproul commented:
bors r-
bors r+

michaelsproul commented:
Ah this is failing because of the web3signer tests and the openssl upgrade. I'll remove it from the current batch and see if I can get it passing by holding back openssl

bors r-

michaelsproul committed:
Regenerate web3signer test certs

The previous certs were causing the tests to fail because they used an old encryption algorithm
which is no longer supported by our bundled version of openssl.

michaelsproul commented:
Ok I think I got it to work without holding any deps. I just regenerated the web3signer certs using `./generate.sh`. This seems to be enough to make the test pass on my machine. Presumably the new version of openssl defaulted to a more modern non-deprecated algorithm.

paulhauner commented:
bors r+

paulhauner commented:
bors retry

jxs commented:
Thanks for the help Michael :)

ralexstokes commented on beacon_node/execution_layer/Cargo.toml:
```diff
@@ -1,56 +1,56 @@
 axum = "0.6"
 hyper = "0.14"
 ethereum-consensus = { git = "https://github.com/ralexstokes/ethereum-consensus", rev = "e380108" }
 ssz_rs = "0.9.0"
```
fwiw I've re-exported the `ssz_rs` `prelude` to `ethereum_consensus::ssz::prelude` if you want to drop this dep
